aries_vcx/src/handlers/connection/inviter/state_machine.rs

```diff
@@ -314,164 +313,119 @@ impl SmConnectionInviter {
                         let problem_report = ProblemReport::create()
                             .set_problem_code(ProblemCode::RequestProcessingError)
                             .set_explain(err.to_string())
-                            .set_thread_id(&state.thread_id);
+                            .set_thread_id(&self.thread_id);
 
-                        send_message(&pairwise_info.pw_vk, &state.did_doc, &problem_report.to_a2a_message()).ok();
+                        (self.send_message)(&self.pairwise_info.pw_vk, &state.did_doc, &problem_report.to_a2a_message()).ok();
                         InviterFullState::Initial((state, problem_report).into())
                     }
                 }
             }
-            _ => state.clone()
+            _ => self.state
         };
-        Ok(Self { state, pairwise_info, ..self })
+        Ok(Self { state, ..self })
     }
 
     pub fn handle_ack(self, ack: Ack) -> VcxResult<Self> {
-        let Self { state, pairwise_info, send_message, .. } = self.clone();
-        let state = match state {
-            InviterFullState::Responded(state) => {
-                let thread_id = self.get_thread_id()?;
-                if !ack.from_thread(&thread_id) {
-                    let problem_report = ProblemReport::create()
-                        .set_problem_code(ProblemCode::RequestProcessingError)
-                        .set_explain(format!("Cannot handle Response: thread id does not match: {:?}", ack.thread))
-                        .set_thread_id(&thread_id); // TODO: Maybe set sender's thread id?
-
-                    send_message(&pairwise_info.pw_vk, &state.did_doc, &problem_report.to_a2a_message()).ok();
-                    InviterFullState::Initial((state, problem_report).into())
-                } else {
-                    InviterFullState::Completed((state, ack).into())
-                }
-            }
-            _ => {
-                state.clone()
-            }
+        let state = match self.state {
+            InviterFullState::Responded(state) => InviterFullState::Completed((state, ack).into()),
+            _ => self.state
         };
-        Ok(Self { state, pairwise_info, ..self })
+        Ok(Self { state, ..self })
     }
 
     pub fn get_thread_id(&self) -> VcxResult<String> {
-        match &self.state {
-            InviterFullState::Invited(state) => state.invitation.get_id(), 
-            InviterFullState::Requested(state) => Ok(state.thread_id.clone()),
-            InviterFullState::Responded(state) => state.signed_response.thread.thid.clone().ok_or(VcxError::from_msg(VcxErrorKind::UnknownError, "Thread ID missing on connection")),
-            InviterFullState::Completed(state) => state.thread_id.clone().ok_or(VcxError::from_msg(VcxErrorKind::UnknownError, "Thread ID missing on connection")),
-            InviterFullState::Initial(_) => Ok(String::new())
-        }
+        Ok(self.thread_id.clone())
     }
 
     fn build_response(
```

Comment:
We are verifying thread id in request conditionally against our invite - instead we can now just verify `self.thread_id` against the thread_id in the request